core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1115,18 +1185,61 @@ class ReplicaManager(val config: KafkaConfig,
     for ((dir, reps) <- replicasByDir) {
       val hwms = reps.map(r => r.partition.topicPartition -> r.highWatermark.messageOffset).toMap
       try {
-        highWatermarkCheckpoints(dir).write(hwms)
+        highWatermarkCheckpoints.get(dir).foreach(_.write(hwms))
       } catch {
         case e: IOException =>
-          fatal("Error writing to highwatermark file: ", e)
-          Exit.halt(1)
+          error("Error writing to highwatermark file: ", e)
+          maybeAddLogFailureEvent(dir)
       }
     }
   }
 
+  def maybeAddLogFailureEvent(logDir: String): Unit = {
+    logDirFailureChannel.maybeAddLogFailureEvent(logDir)
+  }
+
+  def handleLogDirFailure(dir: String) {
+    if (!logManager.isLogDirOnline(dir))
+      return
+
+    info(s"Stopping serving replicas in dir $dir")
+    replicaStateChangeLock synchronized {
+      val newOfflinePartitions = allPartitions.values.filter { partition =>
+        partition.getReplica(config.brokerId) match {
+          case Some(replica) =>
+            replica.log.isDefined && replica.log.get.dir.getParentFile.getAbsolutePath == dir
+          case None => false
+        }
+      }.map(_.topicPartition)
+
+      info(s"Partitions ${newOfflinePartitions.mkString(",")} are offline due to failure on log directory $dir")
+
+      newOfflinePartitions.foreach { topicPartition =>
+        val partition = allPartitions.remove(topicPartition)
```

Comment:
Once a partition is removed from allPartitions, future produce/fetch request will get an UnknownTopicPartitionException, ideally, it seems that they should get a KafkaStorageException for consistency?